gargoyle-postgresql/src/Gargoyle/PostgreSQL.hs

```diff
@@ -74,53 +74,66 @@ getLocalPostgresConnectionString dbDir = do
-    , "-k", absoluteDbDir
-    ] Nothing Nothing
-  fix $ \loop -> do
-    l <- hGetLine err
-    let (tag, rest) = span (/= ':') $ processLog l
-    when (tag == "HINT") loop
-    when (tag /= "LOG") $ fail $ "startLocalPostgres: Unexpected output from postgres: " <> show l
-    when (rest /= ":  database system is ready to accept connections") loop
-  return postgres
+    , "-w"
+    , "-o", escapeMany
+      [ "-h", ""
+      , "-k", absoluteDbDir
+      ]
+    ]) { std_in = NoStream, std_out = UseHandle devNull, std_err = Inherit }
+  r <- waitForProcess postgres
+  case r of
+    ExitSuccess -> return absoluteDbDir
+    _ -> do
+      putStrLn $ "startLocalPostgres failed: " <> show r
+      exitWith r
 
 -- | Perform a "Smart Shutdown" of Postgres;
 -- see http://www.postgresql.org/docs/current/static/server-shutdown.html
-shutdownPostgresSmart :: ProcessHandle -- ^ handle of the PostgreSQL server
+shutdownPostgresSmart :: FilePath -- ^ Path to PostgreSQL `pg_ctl` executable
+                      -> FilePath -- ^ Path where the server to start is located
                       -> IO ()
-shutdownPostgresSmart postgres = do
-  terminateProcess postgres
-  _ <- waitForProcess postgres
-  return ()
+shutdownPostgresSmart = shutdownPostgresWithMode "smart"
 
 -- | Perform a "Fast Shutdown" of Postgres;
 -- see http://www.postgresql.org/docs/current/static/server-shutdown.html
-shutdownPostgresFast :: ProcessHandle -- ^ handle of the PostgreSQL server
-                     -> IO ()
-shutdownPostgresFast postgres = do
-  withProcessHandle postgres $ \p -> do
-    case p of
-      ClosedHandle _ -> return ()
-      OpenHandle h -> signalProcess sigINT h
-  _ <- waitForProcess postgres
-  return ()
+shutdownPostgresFast :: FilePath -- ^ Path to PostgreSQL `pg_ctl` executable
+                      -> FilePath -- ^ Path where the server to start is located
+                      -> IO ()
+shutdownPostgresFast = shutdownPostgresWithMode "fast"
+
+-- | Perform a "Immediate Shutdown" of Postgres;
+-- see http://www.postgresql.org/docs/current/static/server-shutdown.html
+shutdownPostgresImmediate :: FilePath -- ^ Path to PostgreSQL `pg_ctl` executable
+                      -> FilePath -- ^ Path where the server to start is located
+                      -> IO ()
+shutdownPostgresImmediate = shutdownPostgresWithMode "immediate"
+
+shutdownPostgresWithMode :: String -- ^ The shutdown mode to execute; see https://www.postgresql.org/docs/9.5/app-pg-ctl.html
+                         -> FilePath -- ^ Path to PostgreSQL `pg_ctl` executable
+                         -> FilePath -- ^ Path where the server to start is located
+                         -> IO ()
+shutdownPostgresWithMode mode binPath absoluteDbDir = do
+  (_, _, _, postgres) <- createProcess (proc binPath
+    [ "stop"
+    , "-D", absoluteDbDir
+    , "-w"
+    , "-m", mode
+    ]) { std_in = NoStream, std_out = NoStream, std_err = Inherit }
+  r <- waitForProcess postgres
+  case r of
```

Comment:
I think this error reporting should be in a `finally` clause.

Reply:
Same elsewhere

Reply:
Although, with `exitWith` that would also cause executable to exit in the case of an async exception to the thread. Is that what we want?